source/portable/NetworkInterface/MPS3_AN552/NetworkInterface.c

```diff
@@ -0,0 +1,367 @@
+        eEthernetLinkStateTimeOutCount++;
+
+        if( eEthernetLinkStateTimeOutCount >= 5 )
+        {
+            FreeRTOS_debug_printf( ( "Ethernet Link state is DOWN \n " ) );
+            xReturn = pdFAIL;
+        }
+    }
+
+    if( xReturn == pdPASS )
+    {
+        xEthernetLinkInfo = xEthernetInterface0.pxEthernetPHYDriver->GetLinkInfo();
+        ulMACControlArg = xEthernetLinkInfo.speed << ARM_ETH_MAC_SPEED_Pos |
+                          xEthernetLinkInfo.duplex << ARM_ETH_MAC_DUPLEX_Pos |
+                          ARM_ETH_MAC_ADDRESS_BROADCAST;
+        /* Configure Ethernet MAC based on PHY status */
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONFIGURE, ulMACControlArg );
+        /* Enable RX/TX */
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONTROL_RX, 1 );
+        xEthernetInterface0.pxEthernetMACDriver->Control( ARM_ETH_MAC_CONTROL_TX, 1 );
+    }
+
+    return xReturn;
+}
+
+static void prvEthernetDriverNotifications( uint32_t event )
+{
+    BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+
+    configASSERT( xRxTaskHandle );
+
+    if( event == ARM_ETH_MAC_EVENT_RX_FRAME )
+    {
+        /* Ethernet frame received. Send notification to Receive task. */
+        vTaskNotifyGiveFromISR( xRxTaskHandle, &xHigherPriorityTaskWoken );
+    }
+
+    portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
+}
+/*-----------------------------------------------------------*/
+
+static BaseType_t xLAN91C111_NetworkInterfaceOutput( NetworkInterface_t * pxInterface,
+                                                     NetworkBufferDescriptor_t * const pxNetworkBuffer,
+                                                     BaseType_t xReleaseAfterSend )
+{
+    BaseType_t xReturn = pdFAIL;
+
+    ( void ) pxInterface;
+
+    if( pxNetworkBuffer->xDataLength > 1518 )
+    {
+        FreeRTOS_debug_printf( ( "Error: Too big Ethernet frame, Size:", pxNetworkBuffer->xDataLength ) );
+        return xReturn;
+    }
+
+    xReturn = xEthernetInterface0.pxEthernetMACDriver->SendFrame(
+        pxNetworkBuffer->pucEthernetBuffer,
+        pxNetworkBuffer->xDataLength,
+        0 );
+
+    if( xReturn < 0 )
+    {
+        FreeRTOS_debug_printf( ( "Error: Ethernet driver SendFrame returned %d ", xReturn ) );
+    }
+
+    if( xReleaseAfterSend == pdTRUE )
+    {
+        vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+void vNetworkInterfaceAllocateRAMToBuffers( NetworkBufferDescriptor_t pxNetworkBuffers[ ipconfigNUM_NETWORK_BUFFER_DESCRIPTORS ] )
+{
+    /* FIX ME if you want to use BufferAllocation_1.c, which uses statically
+     * allocated network buffers. */
+
+    /* Hard force an assert as this driver cannot be used with BufferAllocation_1.c
+     * without implementing this function. */
+    configASSERT( 0 );
```

Comment:
Yeh good.

Reply:
In my earlier comments, I referred to Xilinx Ultrascale, which does not use phyHandling.c, because it needs a  1GB PHY.

[Here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/0c59e714305dc0eda5178626e680df9a986a29c5/source/portable/NetworkInterface/STM32Hxx/NetworkInterface.c#L626) is an example of how to use phyHandling.c, used by the STM32Hx driver.

Another thing that I didn't mention yet is that the link status should be monitored from time to time. The STM32Hx driver also shows how that can be accomplished.

Reply:
I've created a new function [prvLAN91C111_CheckEthertnetLinkStatus](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1009/files#diff-f3195023098c2a54c04cbf81a4d2d7601e7f02524963671a6f587606695aa2afR146) which is invoked as part of prvRxTask. The prvLAN91C111_CheckEthertnetLinkStatus check the ethernet link status and notifies the IP stack if the link is down. It also sets the `eEMACState` to `xEMAC_Init`, so that the both MAC and PHY drivers are reinitialised when IP stack calls `xLAN91C111_NetworkInterfaceInitialise` on receiving network down event.